In [1]:
import os
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

from tensorflow import keras
from tensorflow.keras.layers import Dense, Conv2D, Flatten, MaxPooling2D, BatchNormalization, Dropout, LSTM, Activation
from tensorflow.keras.models import Sequential
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from tensorflow.keras.utils import to_categorical

In [3]:
data = pd.read_csv('waterQuality.csv')

data = data.drop('ammonia', axis = 1)
data = data.drop('flouride', axis = 1)
data = data.drop('selenium', axis = 1)
data = data.drop('uranium', axis = 1)

data = data[data['is_safe'] != '#NUM!']
data['is_safe'] = data['is_safe'].astype(int)

columns = dict(zip(list(range(data.shape[1])), data.columns.values.tolist()))
is_safe = data['is_safe']

data = pd.DataFrame(StandardScaler().fit_transform(data)).rename(columns = columns)
data['is_safe'] = is_safe
data = data.dropna()
data.head()

aluminium   arsenic    barium   cadmium  chloramine  chromium    copper  \
0   0.777403 -0.480875  1.054205 -0.993241   -0.711941  2.152993 -0.973050   
1   1.306945 -0.599632  1.432447 -1.131948    1.208552  1.598765 -0.223303   
2   0.271572 -0.480875 -0.812340 -0.965499    0.803417  1.044536 -1.202564   
3   0.548198 -0.480875  1.144654 -1.159690    1.968179 -0.802892  1.306793   
4   0.200439 -0.520460 -1.124801 -1.020982    0.191820  1.635713 -0.361012   

   bacteria   viruses      lead  nitrates  nitrites   mercury  perchlorate  \
0 -0.363345 -0.869387 -0.781070  1.129767 -0.348629  0.609178     1.203362   
1  1.002460  0.849782  0.009785 -1.409197  1.046960 -0.739005     0.892977   
2 -0.818614 -0.861452 -0.368450  0.783298 -0.383518  0.272132     1.911763   
3  1.184567  1.008475 -1.434385 -1.517469 -0.069511 -0.401960    -0.415275   
4 -0.575804 -0.866742  0.302058 -0.555658 -0.383518 -0.739005     0.024578   

     radium    silver  is_safe  
0  1.661842  1.338736      1.0  
1  0.124811  0.851134      1.0  
2  1.786699  2.035310      0.0  
3 -0.516695  2.104968      1.0  
4 -0.219622 -0.611672      1.0

In [6]:
X = data.drop('is_safe', axis = 1)
Y = data['is_safe']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.33, random_state = 13)

In [7]:
def build_():
    model = Sequential()
    
    model.add(Dense(256, input_dim = 16))
    model.add(Activation('relu'))
    model.add(Dense(128))
    model.add(Activation('relu'))
    model.add(Dense(128))
    model.add(Activation('relu'))
    model.add(Dense(64))
    model.add(Activation('relu'))
    model.add(Dense(16))
    model.add(Activation('relu'))
    model.add(Dense(2, activation = 'sigmoid'))
    
    model.compile(
        loss = 'categorical_crossentropy',
        optimizer = 'sgd',
        metrics = [keras.metrics.Recall(name = 'recall')]
    )
    return model


class MLP():
    def __init__(self, epochs, batch_size):
        self.clf = KerasClassifier(build_, epochs = epochs, batch_size = batch_size)
        
    def fit_(self, x, y):
        self.clf.fit(x, y, verbose = 0)
    
    def predict_(self, x, y):
        self.p = self.clf.predict(x)
        return np.sum(y == self.p) / len(y)
        
        
clf = MLP(epochs = 250, batch_size = 32)
clf.fit_(X_train, Y_train)

clf.predict_(X_test, Y_test)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:26: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.


0.9249431387414708

92% accuracy is a bit lower than FOLD-R++'s attempt but this is still solid performance for such a basic model. 